src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java

```diff
@@ -25,76 +31,142 @@
+                    .peek(readWithIndex -> {
+                        final GATKRead read = readWithIndex.getValue();
+                        if (!(read.isSecondaryAlignment()||read.isSupplementaryAlignment())) {
+                            PairedEnds fragment = (ReadUtils.readHasMappedMate(read)) ?
+                                    PairedEnds.placeHolder(read, header, readWithIndex.getIndex()) :
+                                    PairedEnds.newFragment(read, header, readWithIndex.getIndex(), scoringStrategy);
+
+                            out.add(new Tuple2<>(fragment.key(header), fragment));
+                        } else {
+                            hadNonPrimaryRead.set(readWithIndex);
+                        }
+                    })
+                    .filter(readWithIndex -> ReadUtils.readHasMappedMate(readWithIndex.getValue()))
 
-        JavaPairRDD<String, Iterable<PairedEnds>> keyedPairs = keyedReads.flatMapToPair(keyedRead -> {
-            List<Tuple2<String, PairedEnds>> out = Lists.newArrayList();
-            // Write each read out as a pair with only the first slot filled
-            for (GATKRead read : keyedRead._2()) {
-                read.setIsDuplicate(false);
-                final PairedEnds pair = PairedEnds.of(read);
-                out.add(new Tuple2<>(pair.keyForFragment(header), pair));
-            }
+            ////// Making The Paired Reads //////
             // Write each paired read with a mapped mate as a pair
-            final List<GATKRead> sorted = Lists.newArrayList(Iterables.filter(keyedRead._2(), read -> ReadUtils.readHasMappedMate(read)));
-            sorted.sort(new GATKOrder(header));
-            PairedEnds pair = null;
-            //Records are sorted, we iterate over them and pair them up.
-            for (final GATKRead record : sorted) {
-                if (pair == null) {                                //first in pair
-                    pair = PairedEnds.of(record);
-                } else {                                           //second in pair
-                    pair.and(record);
-                    out.add(new Tuple2<>(pair.key(header), pair));
-                    pair = null;                                   //back to first
-                }
-            }
-            if (pair != null) {                                    //left over read
+                    .filter(indexPair -> !(indexPair.getValue().isSecondaryAlignment()||indexPair.getValue().isSupplementaryAlignment()))
+                    .collect(Collectors.toList());
+
+            // Mark duplicates cant properly handle templates with more than two reads in a pair
+            if (primaryReads.size()>2) {
+                throw new GATKException(String.format("Readgroup containing read %s has more than two primary reads, this is not valid", primaryReads.get(0).getValue()));
```

Comment:
This should be a UserException.UnsupportedFeature probably instead of a GATKException.  It's possible in a valid bam, it's just not something we allow.  The message is a bit confusing.  Read group is an overloaded term.  Maybe "MarkDuplicatesSpark only supports singleton fragments and pairs.  We found a group with >2 primary reads.  ( %d number of reads).  \list all reads one line at a time. "

Reply:
done